### PR TITLE
Teamwhitelistsupport merge to integ CiscoM31

### PR DESCRIPTION
**Pull request by ritmanda**

### CODE REVIEW

File: pkg/structs/structs.go [hunk 1]
```diff
@@ -148,7 +148,7 @@ type Contact struct {
 	Verified bool   `json:"is_verified"`
 }
 
-//OpenStaxUser is a retrieved and authenticated user from OpenStax Accounts
```
Comment (kumarkowsikt) [inline pkg/structs/structs.go:151]: remove this change as its not needed for teamwhitelist support

File: pkg/providers/openid/openid.go [hunk 2]
```diff
@@ -57,6 +57,29 @@ func (Provider) GetUserInfo(r *http.Request, user *structs.User, customClaims *s
 		log.Error(err)
 		return err
 	}
+	var f interface{}
+	err = json.Unmarshal(data, &f)
+	if err != nil {
+		log.Error("Error unmarshaling claims")
```
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go:63]: log error

File: pkg/providers/openid/openid.go [hunk 3]
```diff
@@ -57,6 +57,29 @@ func (Provider) GetUserInfo(r *http.Request, user *structs.User, customClaims *s
...
+		return err
+	}
+	if cfg.Cfg.TeamWhiteListClaim != "" {
+		log.Infof("TeamWhiteListClaim is %+v", cfg.Cfg.TeamWhiteListClaim)
```
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go:67]: remove or reduce log level for all instances added.

File: pkg/providers/openid/openid.go [hunk 4]
```diff
@@ -57,6 +57,29 @@ func (Provider) GetUserInfo(r *http.Request, user *structs.User, customClaims *s
...
+		m := f.(map[string]interface{})
+		for k := range m {
+			log.Infof("checking claim %s", k)
+			if k == cfg.Cfg.TeamWhiteListClaim {
+				/*for _, membership := range m[k].([]interface{}) {
+					user.TeamMemberships = append(user.TeamMemberships, membership.(string))
+				}*/
+				log.Infof("claim values of %+v is %+v, converted to %+v", k, m[k], m[k].(string))
```
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go:75]: remove commented code, this log and also should we break here on first match

File: pkg/providers/common/common.go [hunk 5]
```diff
@@ -48,6 +48,7 @@ func PrepareTokensAndClient(r *http.Request, ptokens *structs.PTokens, setProvid
 	}
 
 	log.Debugf("ptokens: accessToken length: %d, IdToken length: %d", len(ptokens.PAccessToken), len(ptokens.PIdToken))
+	log.Debugf("%d ----- %d", ptokens.PAccessToken, ptokens.PIdToken)
```
Comment (kumarkowsikt) [inline pkg/providers/common/common.go:51]: remove this log as it has sensitive information

File: handlers/auth.go [hunk 6]
```diff
@@ -139,26 +139,30 @@ func verifyUser(u interface{}) (bool, error) {
...
-		for _, team := range user.TeamMemberships {
-			for _, wl := range cfg.Cfg.TeamWhiteList {
-				if team == wl {
-					log.Debugf("verifyUser: Success! found user.TeamWhiteList in TeamWhiteList: %s for user %s", wl, user.Username)
-					return true, nil
+		log.Info("User not in userwhitelist, checking in teamWhiteList")
+		// TeamWhiteList
+		if len(cfg.Cfg.TeamWhiteList) != 0 {
+			for _, team := range user.TeamMemberships {
+				log.Infof("teammembership %+v", team)
```
Comment (kumarkowsikt) [inline handlers/auth.go:155]: remove logs which are not necessary in prod

File: handlers/logout.go [hunk 7]
```diff
@@ -39,8 +39,8 @@ func LogoutHandler(w http.ResponseWriter, r *http.Request) {
 	var token = ""
 	if claims != nil {
 		token = claims.PIdToken
+		log.Debugf("id token : %+v", token)
```
Comment (kumarkowsikt) [inline handlers/logout.go:42]: remove this log

File: handlers/logout.go [hunk 8]
```diff
@@ -84,6 +84,7 @@ func LogoutHandler(w http.ResponseWriter, r *http.Request) {
 			q.Add("post_logout_redirect_uri", redirectURL)
 		}
 		if token != "" {
+			log.Debug("adding id token")
```
Comment (kumarkowsikt) [inline handlers/logout.go:87]: remove this log

File: handlers/validate.go [hunk 9]
```diff
@@ -66,6 +66,7 @@ func ValidateRequestHandler(w http.ResponseWriter, r *http.Request) {
 		w.Header().Add(cfg.Cfg.Headers.AccessToken, claims.PAccessToken)
 	}
 	if cfg.Cfg.Headers.IDToken != "" && claims.PIdToken != "" {
+		log.Infof("Id token populated %+v", len(claims.PIdToken))
```
Comment (kumarkowsikt) [inline handlers/validate.go:69]: remove this log

File: pkg/providers/openid/openid.go (hunk 4 above)
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go:75]: remove this log and commented code

File: pkg/providers/openid/openid.go [hunk 10]
```diff
@@ -57,6 +57,29 @@ func (Provider) GetUserInfo(r *http.Request, user *structs.User, customClaims *s
...
+				user.TeamMemberships = append(user.TeamMemberships, m[k].(string))
```
Comment (kumarkowsikt) [inline pkg/providers/openid/openid.go:76]: is permission name always sent as string ? should check if type is string before casting and appending ?

File: pkg/cfg/cfg.go [hunk 11]
```diff
@@ -55,6 +56,7 @@ type Config struct {
 	Domains       []string `mapstructure:"domains"`
 	WhiteList     []string `mapstructure:"whitelist"`
 	TeamWhiteList []string `mapstructure:"teamWhitelist"`
+	TeamWhiteListClaim string   `mapstructure:"teamWhitelistclaim"`
```
Comment (kumarkowsikt) [inline pkg/cfg/cfg.go:59]: format the struct

File: handlers/auth.go [hunk 12]
```diff
@@ -139,26 +139,30 @@ func verifyUser(u interface{}) (bool, error) {
...
+				for _, wl := range cfg.Cfg.TeamWhiteList {
+					if team == wl {
+						log.Debugf("verifyUser: Success! found user.TeamWhiteList in TeamWhiteList: %s for user %s", wl, user.Username)
+						return true, nil
```
Comment (kumarkowsikt) [inline handlers/auth.go:159]: add a comment saying first match will be used among team whitelists

File: handlers/auth.go (hunk 6 above)
Comment (ritmanda) [reply handlers/auth.go:155]: removed this log log.Infof("teammembership %+v", team)

**Pull request merged**